[GSoC] Added the dsolve_system in systems.py

mijo2 opened the pull request:
#### References to other Issues or PRs


#### Brief description of what is fixed or changed
Added `dsolve_system` that can solve system of ODEs passed. For now, only linear first-order systems of ODEs are handled by this function.

#### Other comments
This function can handle a lot of solutions together and many of the helper functions created provide a framework for adding other techniques and functionalities later.


#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* solvers
  * Added `dsolve_system` that can solve a system of ODEs
<!-- END RELEASE NOTES -->

mijo2 committed:
solvers.ode: Added the first version of dsolve_system

mijo2 committed:
solvers.ode: Added an upgraded version of dsolve_system

The first commit added the dsolve_system but there where lots of
logical errors in that version. This version has been tested with
some examples of an explicit type 1 systems of ODEs and its working
well on these systems.

mijo2 committed:
solvers.ode: Incorporating dsolve_system in dsolve

mijo2 commented:
@oscarbenjamin If there an implicit system like below:
```
⎡          2                              ⎤
⎢⎛d       ⎞     2             d           ⎥
⎢⎜──(f(t))⎟  = g (t), -f(t) + ──(g(t)) = 0⎥
⎣⎝dt      ⎠                   dt          ⎦

```
Then should we have different integration constants for both of them in the solution? Like using `C1` and `C2` for one linear system and `C3` and `C4` for another.

mijo2 committed:
solvers.ode: Updated the dsolve_system

Updated it to make it work for multiple systems and added a piece of
code that can get the functions out of the system if the functions
are not passed

oscarbenjamin commented on sympy/solvers/ode/systems.py:
```diff
@@ -1099,40 +1110,64 @@ def _ode_component_solver(eqs, funcs, t, const_idx=0):
-    sol = []
+    # Note: This is added to solve a major problem encountered.
+    # Might be best to make a function for this function
+    # extraction later.
```
What is the major problem?

mijo2 replied:
Well, I wanted to add `dsolve_system` call in `dsolve` itself. But if I did that, then there is a high chance that a user will not pass the `funcs` which would make us depend on `classify_sysode`. Now, `classify_sysode` calls our matching function and returns its output. Since our `dsolve_system` is made so that even an end-user can call it, we are not passing the match duct. This would mean we would have to call our matching function twice. 

The above was the problem. To solve that, I simply added this `funcs` extractor code. This removed the risk of calling the matching function twice.

oscarbenjamin replied:
It's probably worth making this code for guessing the functions into a separate function that can be used in both places.

oscarbenjamin commented on sympy/solvers/ode/systems.py:
```diff
@@ -1064,3 +1066,153 @@ def neq_nth_linear_constant_coeff_match(eqs, funcs, t):
+    return None
+
+
+def dsolve_system(eqs, funcs=None, t=None):
```
I think it's better to require that the user pass all arguments like `funcs` and `t`. A user who doesn't want to supply those can call `dsolve`.

mijo2 replied:
As I said in the previous comment, in order to do that, it would make calling `dsolve_system` from `dsolve` very inefficient.

oscarbenjamin commented on sympy/solvers/ode/systems.py:
```diff
@@ -1064,3 +1066,153 @@ def neq_nth_linear_constant_coeff_match(eqs, funcs, t):
+    not_solved_systems = []
+
+    for wcc in components:
+        loop_sol = []
```
This loop is quite complicated and I don't understand what it is doing. I think it's probably better to refactor this with one or more functions somehow.

oscarbenjamin replied:
There should probably be a function for solving a single weakly connected component.

mijo2 replied:
@oscarbenjamin Sure thing, I would add such a function with proper comments explaining the blocks of code, as this loop is complicated for me as well. This became complicated since in the last PR, we introduced the fact that our systems solver method can get multiple solutions for an implicit system. Incorporating this with the plan of weakly and strongly connected components made the design lot more complicated.

mijo2 committed:
solvers.ode: Made a weakly connected component solver function

Added this function and properly explained the loop to solve wcc
with comments

mijo2 committed:
solvers.ode: Added the test cases properly

Updated some of the test cases

mijo2 committed:
solvers.ode: Updated test cases in test_ode.py and funcs extractor

mijo2 committed:
solvers.ode: Added ics solving in dsolve_system

mijo2 committed:
solvers.ode, doc: Added documentation for dsolve_system

mijo2 committed:
solvers.ode: Added more test cases and fixed the output

Added more test cases for better coverage and fixed the output
of dsolve_system when the function isn't able to solve the whole
system.

mijo2 commented:
@oscarbenjamin Except one minor error, most of this PR is done. Regarding your point about making `funcs` and `t` as required arguments, I don't know for now how to do it without risking to call `classify_sysode` from `dsolve` since a user passing a system of ODEs won't always pass the required arguments.

mijo2 committed:
solvers.ode: Some minor changes

1. Added a special condition for ics
2. Added more information about ics in docstring
3. Added test cases for better coverage
4. Fixed a minor error in the docstring

oscarbenjamin commented on sympy/solvers/ode/systems.py:
```diff
@@ -1064,3 +1068,323 @@ def neq_nth_linear_constant_coeff_match(eqs, funcs, t):
+
+            # Note: This const_idx logic has to be
+            # verified
+            return sols, temp_const_idx
```
I don't understand what `const_idx` is doing. I would write this like
```
if match.get('is_implicit', False):
    canon_eqs = match['canon_eqs']
else
    canon_eqs = [eqs]
for canon_eq in canon_eqs:
    if is_linear(canon_eq)...
```
Or something like that...

mijo2 replied:
`const_idx` keeps tracks of the constant naming across the sub-systems of ODEs since we don't want the sub-system solutions to that same constants in their solutions.

> Or something like that...

Sure, I would do that soon.

oscarbenjamin commented on sympy/solvers/ode/systems.py:
```diff
@@ -1064,3 +1068,323 @@ def neq_nth_linear_constant_coeff_match(eqs, funcs, t):
+
+        # If this is the first SCC to be solved,
+        # then looping isn't necessary.
+        if not loop_sol:
```
The flow control in this function is very confusing. I think there is a cleaner way to write this without this variable.

mijo2 replied:
Ok, I would try to write this in cleaner and most understandable manner.

mijo2 replied:
I was able to make this function more compact and cleaner but I wasn't able to eliminate `loop_sol` variable, since it is the variable that stores the solutions.

mijo2 committed:
solvers.ode: Made some changes

Simplified the following functions:
1. _weak_component_solver
2. _ode_component_solver

mijo2 commented:
@oscarbenjamin I don't understand why codecov is showing me some lines are untested while the majority of those lines are in fact tested. Mainly, it is showing that the exceptions haven't been checked properly but I have added at least one test case for every exception that I have added.

oscarbenjamin commented:
I don't think that the codecov coverage report updates when you push more commits so it shows the coverage from the first time that the tests all passed in the PR.

Also it only measures coverage from the not-slow tests. Ideally the tests in the ODE module would be separated into a minimal set of fast tests that has 100% line coverage and then the rest of the tests which would all be marked as slow.

mijo2 commented:
@oscarbenjamin I have almost added tests for every line and none of them are slow. Then it's possible that codecov is showing the results by merging this branch with the master one. That would explain why the coverage has dropped so suddenly compared to my previous update. Which would mean something in the master branch isn't tested properly.

mijo2 committed:
solvers.ode: Rearranged test cases

For better coverage. This is an experimental commit. Will be
squashed later.

mijo2 commented:
@oscarbenjamin I have started working on the framework as discussed, but before I work any further, I just wanted to confirm one fact, if a first-order system of ODEs is such that, the first-order terms are linear, then will that system have only one solution(assuming it exists), even if the dependent variables(0 order terms) are non-linear?

mijo2 committed:
solvers.ode: Simplified dsolve_system

Added the weakly and strongly connected component solver functions
separately, outside of dsolve_system.

oscarbenjamin commented:
I'm not sure I understand the question but in general the solutions of initial value problems are usually unique:
https://en.wikipedia.org/wiki/Picard%E2%80%93Lindel%C3%B6f_theorem

There can still be multiple forms for the general solution though that would be valid for different initial conditions.

mijo2 commented:
> There can still be multiple forms for the general solution though that would be valid for different initial conditions.

Yes, so with this theorem, it is safe to assume that if we pass a system of first-order ODEs in its canonical form(output from canonical_odes), then it would have only one solution. I have used this logic to further greatly simplify the component solvers.

mijo2 committed:
solvers.ode: Replaced the constant naming with dummies

mijo2 commented on sympy/solvers/ode/systems.py:
```diff
@@ -1064,3 +1074,263 @@ def neq_nth_linear_constant_coeff_match(eqs, funcs, t):
+                       key=lambda x: x.dummy_index)
+    dummy_to_constants = {c: Symbol('C{}'.format(i+1)) for i, c, in enumerate(constants)}
+    sol = [s.subs(dummy_to_constants) for s in sol]
+
```
I am using this kind of method to replace the dummies made with the constants. Is this a viable approach?

oscarbenjamin replied:
Yes, but is there not already a function to handle this in `ode.py`?

mijo2 replied:
I don't think so. I tried finding such function but wasn't able to.

mijo2 replied:
The closest to the required function is `constant_renumber` function but I don't know how to use it for a list of equations. Since, when the input is:
```
⎡                x                  x⎤
⎣f(x) = Dummy₂₈⋅ℯ , g(x) = Dummy₂₉⋅ℯ ⎦

```
The output is:
```
⎡           x             x⎤
⎣f(x) = C₁⋅ℯ , g(x) = C₁⋅ℯ ⎦
```

oscarbenjamin replied:
Maybe `constant_renumber` should be fixed.

mijo2 replied:
Should I fix it in this PR?

oscarbenjamin replied:
It can be done later.

mijo2 committed:
solvers.ode: Some changes

Fixed docstring explanation and  test cases, added a doit option
in dsolve_system and removed an unnecessary key value pair

mijo2 commented:
I have simplified the `_weak_component_solver` and `strong_component_solver` and removed the constant number tracking. After the test cases have passed, maybe we can get this merged(if everything looks fine) and work on implementing component division.

oscarbenjamin commented:
Looks like the docs build has failed

mijo2 committed:
solvers.ode: Made a separate private function

Made a separate private function for extracting dependent variables
from a system of ODEs

mijo2 commented:
@oscarbenjamin The build has passed

oscarbenjamin commented:
Looks good.
